src/Database/Franz/Network.hs

```diff
@@ -171,72 +181,95 @@ getResponse = do
                 return vres
 
 -- | Fetch requested data from the server.
--- Termination of the continuation cancels the request, allowing flexible
--- control of its lifetime.
+--
+-- Termination of 'fetch' continuation cancels the request, allowing
+-- flexible control of its lifetime.
 fetch :: Connection
   -> Query
-  -> (STM Response -> IO r) -- ^ running the STM action blocks until the response arrives
+  -> (Contents -> IO r)
+  -- ^ Action to run on results that are available instantly.
+  -> Either (IO r) (IO (Contents -> IO r))
+  -- ^ Action to run if results are delayed. If we don't want to wait
+  -- for delayed results, use 'Left' as a default action. 'Right'
+  -- action is ran on delayed result when it becomes available.
```

Comment:
This seems to make sense but you can no longer compose `fetch`es easily using `Codensity`...

Reply:
Can you give an example? Maybe we can improve this API.

Reply:
The old API allowed us to write
```haskell
do
  t0 <- Codensity $ fetch conn q0 
  t1 <- Codensity $ fetch conn q1
  ...
```
of course with traversing containers or HKD records as well.

Reply:
I don't understand what Codensity does in this case. I'll read the paper linked in Hackage but if you can explain simply what this actually does, that'd be helpful.

Reply:
It's the same as ContT but without an explicit type parameter: it's a way to chain `with-` style function sequentially, and `flip runCodensity return` runs the cleanups in the reverse order.

Reply:
The function is no longer in `with` style so it doesn't seem necessary, right? You can just write this:

```haskell
fetchSequential :: Connection -> IO ()
fetchSequential conn = do
  t0 <- fetchBlocking q0
  t1 <- fetchBlocking q1
  t2 <- fetchBlocking q2
  print $ t0 <> t1 <> t2
  where
    q0, q1, q2 :: Query
    (q0,q1,q2) = undefined

    fetchBlocking :: Query -> IO Contents
    fetchBlocking q = fetch conn q pure (Right $ pure pure)
```

If you need clean-ups, you just use usual IO tools like onException &c.

I think if you really want, you can recover the old type signature, no? Something like this, can be written in user-land.

```haskell
fetchStm
  :: Connection
  -> Query
  -> (STM Response -> IO r)
  -> IO r
fetchStm = \conn q cont -> do
  resVar <- newEmptyTMVarIO
  let onInstant = atomically . putTMVar resVar . Left
      onDelayed = do
        res' <- newEmptyTMVarIO
        atomically . putTMVar resVar . Right $ takeTMVar res'
        pure $ atomically . void . putTMVar res'
  -- Let the async run even after `cont` terminates such that it's not
  -- killed early.
  t <- async $ fetch conn q onInstant (Right onDelayed)
  -- Checks that the thread didn't die with an exception which could
  -- result in the result never being filled. If it did die with an
  -- exception, re-throw it in user's continuation.
  let checkAlive = pollSTM t >>= \case
        -- Thread still running.
        Nothing -> retry
        -- Thread died with exception, re-throw.
        Just (Left e) -> throwSTM e
        -- Thread finished successfully. Presumably the result TMVar
        -- should be filled now, simply retry.
        Just Right{} -> retry

  -- We use readTMVar instead of takeTMVar: if user re-runs the
  -- continuation, it will produce same result rather than dead-lock.
  cont $ readTMVar resVar `orElse` checkAlive
```

Reply:
The reason why it's continuation passing style with STM is that users can easily implement behaviour that cancels requests. If the intended way of using new `fetch` is not with-style it's not obvious how I'm supposed to use; I'm not a big fan of design that forces users to use `async` for each request either.

Reply:
>If the intended way of using new fetch is not with-style it's not obvious how I'm supposed to use;

You just use it in straight-forward way... Why do you need `with`-style?

The only things you can do with (STM Response -> IO r) are run it or not run it. You can interrupt it mid-way too. You can do all these things with new way too, right?

I'm simply missing the use-case of previous approach. What can you do with it that the current way is lacking?

> I'm not a big fan of design that forces users to use async for each request either.

`fetch` doesn't force anything, it's just a building block. From it, you can build convenient things like fully-blocking fetches, only instant fetches, asynchronous delayed fetches, whatever your usecase...

I think I should go find how it's used in trader to see it in action.

Reply:
As far as I can see from the code, the intended way to cancel a query while waiting for response is to throw an exception from onDelayed'e and catch an exception outside? That's much less convenient right? Sure, you can define a wrapper function but using `async` also has a considerable overhead no? I'm not convinced that this is simplification.

Reply:
We have discussed internally and decided to go back to the old type albeit with new implementation, one that throws an exception if the callback action is executed outside of fetch block which was my main concern. I still don't like it but I think at least request cancelling is better now and if the user does run things concurrently (actual concurrency with threads) then the contention should be a lot lower on the global lock.

PTAL

Reply:
Building apps on top of this branch